packages/teleport/src/DesktopSession/DesktopSession.tsx

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React from 'react';
+import styled from 'styled-components';
+import useDesktopSession, { State } from './useDesktopSession';
+
+export default function Container() {
+  const state = useDesktopSession();
+  return <DesktopSession {...state} />;
+}
+
+export function DesktopSession(props: State) {
+  const { attempt, setAttempt, tdpClientRef, tdpClientInitialized } = props;
+  const canvasRef = React.useRef<HTMLCanvasElement>(null);
+
+  React.useEffect(() => {
+    resizeCanvasToDisplaySize(canvasRef.current);
+  }, []);
+
+  // Waits for the state hook to initialize the TdpClient.
+  // Once the client is initialized, calls for the client to connect to the
+  // server and passes in the canvas for the client to render to.
+  React.useEffect(() => {
+    setAttempt({ status: 'processing' });
+    const tdpClient = tdpClientRef.current;
+    if (tdpClientInitialized) {
+      try {
+        tdpClient.connect(canvasRef.current);
+        setAttempt({ status: 'success' });
+      } catch (err) {
+        setAttempt({ status: 'failed', statusText: err });
+      }
+    }
+  }, [tdpClientInitialized]);
+
+  // Canvas has two size attributes: the dimension of the pixels in the canvas (canvas.width)
+  // and the display size of the html element (canvas.clientWidth). resizeCanvasToDisplaySize
+  // ensures the two remain equal.
+  function resizeCanvasToDisplaySize(canvas: HTMLCanvasElement) {
+    // look up the size the canvas is being displayed
+    const width = canvas.clientWidth;
+    const height = canvas.clientHeight;
+
+    // If it's resolution does not match change it
+    if (canvas.width !== width || canvas.height !== height) {
+      canvas.width = width;
+      canvas.height = height;
+    }
+  }
+
+  return (
+    <StyledDesktopSession>
+      <canvas ref={canvasRef} />
+    </StyledDesktopSession>
+  );
+}
```

Comment:
Lets decouple the canvas renderer functions from the client. You can use EventEmitter to emit and subscribe to `tdp` events in our component

Something like this:

```tsx
import Codec, { MessageType } from './codec';
import Logger from 'shared/libs/logger';
import { EventEmitter } from 'events';
const logger = Logger.create('TDPClient');
export default class Client extends EventEmitter {
  socketAddr: string;
  username: string;
  codec: Codec;
  socket: WebSocket;

  constructor(socketAddr: string, username: string) {
    super()
    this.socketAddr = socketAddr;
    this.username = username;
    this.codec = new Codec();
  }

  // Pass the canvas to draw to on connection
  connect() {
    try {
      this.socket = new WebSocket(this.socketAddr);
      this.socket.onopen = () => {
        this.socket.send(this.codec.encodeUsername(this.username));
        this.emit('open');
        logger.info('websocket is open');
      };

      this.socket.onmessage = (ev: MessageEvent) => {
        this.processMessage(ev);
      };

      this.socket.onclose = () => {
        this.emit('close')
        logger.info('websocket is closed');
      };

      this.socket.onerror = () => {
        this.handleError(new Error('websocket internal error'));
      };

    } catch (err) {
      this.handleError(err);
    }
  }

  processMessage(ev: MessageEvent) {
    this.codec
      .decodeMessageType(ev.data)
      .then(messageType => {
        if (messageType === MessageType.PNG_FRAME) {          )
          this.processFrame(ev.data);
        } else {
          this.handleError(new Error(`unsupported message type: ${messageType}`))
        }
      })
      .catch(this.handleError);
  }

  processFrame(blob: Blob) {
    Promise.all([
      this.codec.decodeRegion(blob),
      this.codec.decodePng(blob)])
      .then(values => {
        const { left, top } = values[0];
        const bitmap = values[1];
        this.emit('render', { bitmap, left, top })
      })
      .catch(this.handleError);
  }

  resize(w: number, h: number) {
    this.socket.send(this.codec.encodeScreenSpec(w, h));
  }

  handleError(err: Error) {
    this.emit('error', err)
    logger.error(msg, err);
  }
}
```

And then you can subscribe to it's events in our component

```tsx
export function DesktopSession({ addr, username }: State) {  
  const { attempt, setAttempt } = useAttempt('processing');
  const canvasRef = React.useRef<HTMLCanvasElement>(null);

  React.useEffect(() => {
    const tdpClient = new TdpClient({ addr, username });
    const canvas = canvasRef.current;

    fitScreenSize(canvas);

    tdpClient.on('render', data => {
      const { left, top, bitmap } = data;
      canvas.getContext('2d').drawImage(bitmap, left, top);
    });

    tdpClient.on('open', () => {
      setAttempt({ status: 'success' });
      tdpClient.resize(canvas.width, canvas.height);
    });

    tdpClient.on('error', (err: Error) => {
      setAttempt({ status: 'failed', statusText: err.message });
    });

    tdpClient.on('closed', () => {});

    return () => {
      tdpClient.removeAllListeners();
      tdpClient.disconnect();
    };
  }, [addr, username]);

  return (
    <StyledDesktopSession>
      <canvas ref={canvasRef} />
    </StyledDesktopSession>
  );
}
```

Reply:
Got it. In light of converting this to an `Emitter`, do you have any thoughts/experience from working with [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers)? 

They have a similar event driven architecture, and I was thinking about running the client in a worker. Performance is extremely important for this component and right now the draws are extremely slow and noticeable to the naked eye, so I was thinking moving this into its own thread might help mitigate that.

Reply:
Have you tried gathering any performance metrics to see where the bottleneck is? Yes we can discuss.